src/ore/src/vec.rs

```diff
@@ -92,6 +92,11 @@ pub fn swap_remove_multiple<T>(v: &mut Vec<T>, mut indexes: Vec<usize>) -> Vec<T
     result
 }
 
+/// Create a new vector by cloning the elements of the given vector.
+pub fn clone_elements<T: Clone>(vec: &Vec<&T>) -> Vec<T> {
+    vec.iter().map(|c| (*c).clone()).collect::<Vec<_>>()
+}
```

Comment:
If I do this change I get:
```
error[E0277]: a value of type `Vec<T>` cannot be built from an iterator over elements of type `&T`
    --> src/ore/src/vec.rs:98:25
     |
98   |     vec.iter().cloned().collect::<Vec<T>>()
     |                         ^^^^^^^ value of type `Vec<T>` cannot be built from `std::iter::Iterator<Item=&T>`
     |
     = help: the trait `FromIterator<&T>` is not implemented for `Vec<T>`
     = help: the trait `FromIterator<T>` is implemented for `Vec<T>`
```
So it seems that the compiler somehow decides to clone the references instead of the things that they are referring to? But I don't understand why this is happening, because usually when I call `clone` on a reference then it gets called on the thing that it's referring to.

Reply:
I noticed that this works:
```
pub fn clone_elements<T: Clone>(vec: Vec<&T>) -> Vec<T> {
    vec.into_iter().cloned().collect::<Vec<_>>()
}
```
but not this:
```
pub fn clone_elements<T: Clone>(vec: &Vec<&T>) -> Vec<T> {
    vec.into_iter().cloned().collect::<Vec<_>>()
}
```
or this:
```
pub fn clone_elements<T: Clone>(vec: Vec<&T>) -> Vec<T> {
    vec.iter().cloned().collect::<Vec<_>>()
}
```

I strongly suspect that the reason is that if you call iter() on `Vec<&T>`, the elements become type `&&T`, so when you call `cloned()`, it only converts them to `&T`.

Reply:
Hm, ok, makes sense, thanks! I changed it to consume the vector.